### ISSUE TITLE
permissions to use recursive parameter

### ISSUE THREAD

**Issue by navi86** [GoogleCloudPlatform/gcr-cleaner#80]
I have created sa using
```
gcloud artifacts repositories add-iam-policy-binding "${REPO_NAME}" \
    --project "${PROJECT_ID}" \
    --location "${LOCATION}" \
    --member "serviceAccount:gcr-cleaner@${PROJECT_ID}.iam.gserviceaccount.com" \
    --role "roles/artifactregistry.repoAdmin"
```
however, if I try to remove docker images and provide only path to repo `"europe-west2-docker.pkg.dev/**project_id**/python"` with parameter "recursive" it returns nothing, however if I grant sa "gcr-cleaner" role `"roles/storage.objectViewer"` or permission "resourcemanager.projects.get" then it works.
Is it by design ? how is it possible to limit access to this user ?

**Comment by sethvargo** [comment]
Hi @navi86 

Are you using Google Container Registry or Google Artifact Registry? If you are using Google Container Registry, you must grant permissions on the Cloud Storage bucket. If you are using Artifact Registry, Cloud Storage permissions are irrelevant.

**Comment by navi86** [comment]
@sethvargo I am using artifact registry, but without that role/permission, I am not able to provide `"europe-west2-docker.pkg.dev/**project_id**/python"` as repo without specifying image name. The flag "recursive" is simply doesn't work.

**Comment by sethvargo** [comment]
Can you make sure you're using the latest version of gcr-cleaner and run the command with the environment variable `GCRCLEANER_LOG=debug` and paste the results here please?

**Comment by navi86** [comment]
@sethvargo 
one more thing, flag "recursive" with "dry-run" doesn't work. I've run 
`{"dry-run":true,"recursive":true,"repos":["europe-west2-docker.pkg.dev/secure-theme-345716/alpine","europe-west2-docker.pkg.dev/secure-theme-345716/python"],"tag_filter_all":"[a-k,w-z]|[\\d]+?v"}% `
this code removed all docker images in repo base on filter.

**Comment by navi86** [comment]
@sethvargo 
this is log when I run without permissions/role that I mentioned before
`[
    {
    "insertId": "508onrf67g7y9",
    "jsonPayload": {
      "targetType": "HTTP",
      "@type": "type.googleapis.com/google.cloud.scheduler.logging.AttemptStarted",
      "jobName": "projects/project_id/locations/europe-west2/jobs/artifactory-cleaner",
      "url": "https://artifactory-cleaner-yu6tjkdcma-nw.a.run.app/http",
      "scheduledTime": "2022-04-11T08:00:00.541607Z"
    },
    "resource": {
      "type": "cloud_scheduler_job",
      "labels": {
        "location": "europe-west2",
        "project_id": "project_id",
        "job_id": "artifactory-cleaner"
      }
    },
    "timestamp": "2022-04-08T20:15:35.734640292Z",
    "severity": "INFO",
    "logName": "projects/project_id/logs/cloudscheduler.googleapis.com%2Fexecutions",
    "receiveTimestamp": "2022-04-08T20:15:35.734640292Z"
  },
  {
    "insertId": "625097e8000540396f07085f",
    "jsonPayload": {
      "since": "2022-04-08T20:15:35.79369234Z",
      "message": "deleting refs",
      "repos": [
        "europe-west2-docker.pkg.dev/project_id/alpine",
        "europe-west2-docker.pkg.dev/project_id/python"
      ]
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "location": "europe-west2",
        "configuration_name": "artifactory-cleaner",
        "service_name": "artifactory-cleaner",
        "revision_name": "artifactory-cleaner-4jkng",
        "project_id": "project_id"
      }
    },
    "timestamp": "2022-04-08T20:15:36Z",
    "severity": "INFO",
    "labels": {
      "instanceId": "00bf4bf02da6c0c10a67e9ff0d8d6ca1262405effc466c56c9c1bdd67fdd547a6390077a19d8d0a0103ab75d3f79b5e1f5119e3e5043fa21b21cb404605c3254"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:36.348166337Z"
  },
  {
    "insertId": "625097e8000540579285c8cc",
    "jsonPayload": {
      "repo": "europe-west2-docker.pkg.dev/project_id/alpine",
      "message": "deleting refs for repo"
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "configuration_name": "artifactory-cleaner",
        "location": "europe-west2",
        "service_name": "artifactory-cleaner",
        "project_id": "project_id",
        "revision_name": "artifactory-cleaner-4jkng"
      }
    },
    "timestamp": "2022-04-08T20:15:36Z",
    "severity": "INFO",
    "labels": {
      "instanceId": "00bf4bf02da6c0c10a67e9ff0d8d6ca1262405effc466c56c9c1bdd67fdd547a6390077a19d8d0a0103ab75d3f79b5e1f5119e3e5043fa21b21cb404605c3254"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:36.348166337Z"
  },
  {
    "insertId": "625097e800054061c33843e7",
    "jsonPayload": {
      "repo": "europe-west2-docker.pkg.dev/project_id/alpine",
      "message": "computed repo"
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "location": "europe-west2",
        "service_name": "artifactory-cleaner",
        "configuration_name": "artifactory-cleaner",
        "project_id": "project_id",
        "revision_name": "artifactory-cleaner-4jkng"
      }
    },
    "timestamp": "2022-04-08T20:15:36Z",
    "severity": "DEBUG",
    "labels": {
      "instanceId": "00bf4bf02da6c0c10a67e9ff0d8d6ca1262405effc466c56c9c1bdd67fdd547a6390077a19d8d0a0103ab75d3f79b5e1f5119e3e5043fa21b21cb404605c3254"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:36.348166337Z"
  },
  {
    "insertId": "625097e80007c31e42abbbd5",
    "jsonPayload": {
      "repo": "europe-west2-docker.pkg.dev/project_id/python",
      "message": "deleting refs for repo"
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "location": "europe-west2",
        "project_id": "project_id",
        "revision_name": "artifactory-cleaner-4jkng",
        "configuration_name": "artifactory-cleaner",
        "service_name": "artifactory-cleaner"
      }
    },
    "timestamp": "2022-04-08T20:15:36Z",
    "severity": "INFO",
    "labels": {
      "instanceId": "00bf4bf02da6c0c10a67e9ff0d8d6ca1262405effc466c56c9c1bdd67fdd547a6390077a19d8d0a0103ab75d3f79b5e1f5119e3e5043fa21b21cb404605c3254"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:36.681678680Z"
  },
  {
    "insertId": "625097e80007c334fa840668",
    "jsonPayload": {
      "message": "computed repo",
      "repo": "europe-west2-docker.pkg.dev/project_id/python"
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "service_name": "artifactory-cleaner",
        "revision_name": "artifactory-cleaner-4jkng",
        "location": "europe-west2",
        "configuration_name": "artifactory-cleaner",
        "project_id": "project_id"
      }
    },
    "timestamp": "2022-04-08T20:15:36Z",
    "severity": "DEBUG",
    "labels": {
      "instanceId": "00bf4bf02da6c0c10a67e9ff0d8d6ca1262405effc466c56c9c1bdd67fdd547a6390077a19d8d0a0103ab75d3f79b5e1f5119e3e5043fa21b21cb404605c3254"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:36.681678680Z"
  },
  {
    "insertId": "625097e80009a8783556d5bc",
    "jsonPayload": {
      "refs": {},
      "message": "deleted refs"
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "revision_name": "artifactory-cleaner-4jkng",
        "location": "europe-west2",
        "project_id": "project_id",
        "service_name": "artifactory-cleaner",
        "configuration_name": "artifactory-cleaner"
      }
    },
    "timestamp": "2022-04-08T20:15:36Z",
    "severity": "INFO",
    "labels": {
      "instanceId": "00bf4bf02da6c0c10a67e9ff0d8d6ca1262405effc466c56c9c1bdd67fdd547a6390077a19d8d0a0103ab75d3f79b5e1f5119e3e5043fa21b21cb404605c3254"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:36.681678680Z"
  },
  {
    "insertId": "625097e80009ae2787315953",
    "httpRequest": {
      "requestMethod": "POST",
      "requestUrl": "https://artifactory-cleaner-yu6tjkdcma-nw.a.run.app/http",
      "requestSize": "1434",
      "status": 200,
      "responseSize": "744",
      "userAgent": "Google-Cloud-Scheduler",
      "remoteIp": "107.178.196.240",
      "serverIp": "216.239.38.53",
      "latency": "0.843617402s",
      "protocol": "HTTP/1.1"
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "configuration_name": "artifactory-cleaner",
        "project_id": "project_id",
        "location": "europe-west2",
        "revision_name": "artifactory-cleaner-4jkng",
        "service_name": "artifactory-cleaner"
      }
    },
    "timestamp": "2022-04-08T20:15:36.634407Z",
    "severity": "INFO",
    "labels": {
      "instanceId": "00bf4bf02da6c0c10a67e9ff0d8d6ca1262405effc466c56c9c1bdd67fdd547a6390077a19d8d0a0103ab75d3f79b5e1f5119e3e5043fa21b21cb404605c3254"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Frequests",
    "trace": "projects/project_id/traces/50d9b977c31f9446ec036752a95864a4",
    "receiveTimestamp": "2022-04-08T20:15:36.640531243Z"
  },
  {
    "insertId": "2s42pyf4sqn94",
    "jsonPayload": {
      "jobName": "projects/project_id/locations/europe-west2/jobs/artifactory-cleaner",
      "@type": "type.googleapis.com/google.cloud.scheduler.logging.AttemptFinished",
      "url": "https://artifactory-cleaner-yu6tjkdcma-nw.a.run.app/http",
      "targetType": "HTTP"
    },
    "httpRequest": {
      "status": 200
    },
    "resource": {
      "type": "cloud_scheduler_job",
      "labels": {
        "location": "europe-west2",
        "project_id": "project_id",
        "job_id": "artifactory-cleaner"
      }
    },
    "timestamp": "2022-04-08T20:15:36.648556785Z",
    "severity": "INFO",
    "logName": "projects/project_id/logs/cloudscheduler.googleapis.com%2Fexecutions",
    "receiveTimestamp": "2022-04-08T20:15:36.648556785Z"
  },
  {
    "textPayload": "using default token resolution for authentication",
    "insertId": "625097ee0007c4e66c3d2869",
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "configuration_name": "artifactory-cleaner",
        "project_id": "project_id",
        "service_name": "artifactory-cleaner",
        "revision_name": "artifactory-cleaner-4jkng",
        "location": "europe-west2"
      }
    },
    "timestamp": "2022-04-08T20:15:42Z",
    "severity": "DEBUG",
    "labels": {
      "instanceId": "00bf4bf02dc69706490e380fabefbe1a725ee9e5fb08a68cc818b5bdf1cb0febf175c5670785f5e48ff1a1b7e764b56e659649affb6154a0817bd2f2da81b8"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:42.511719307Z"
  },
  {
    "insertId": "625097ee0007f2ded9ab1a7e",
    "jsonPayload": {
      "message": "server is listening",
      "port": "8080"
    },
    "resource": {
      "type": "cloud_run_revision",
      "labels": {
        "service_name": "artifactory-cleaner",
        "revision_name": "artifactory-cleaner-4jkng",
        "project_id": "project_id",
        "location": "europe-west2",
        "configuration_name": "artifactory-cleaner"
      }
    },
    "timestamp": "2022-04-08T20:15:42Z",
    "severity": "INFO",
    "labels": {
      "instanceId": "00bf4bf02dc69706490e380fabefbe1a725ee9e5fb08a68cc818b5bdf1cb0febf175c5670785f5e48ff1a1b7e764b56e659649affb6154a0817bd2f2da81b8"
    },
    "logName": "projects/project_id/logs/run.googleapis.com%2Fstderr",
    "receiveTimestamp": "2022-04-08T20:15:42.846736383Z"
  }
]`
so basically there is no child repo

**Comment by sethvargo** [comment]
@navi86 the JSON payload is is `dry_run`, not `dry-run`.

Can you try doing this with the CLI locally? It's very difficult to parse the JSON logs because they include values from other services and are out of order. The second log shows nothing was a candidate to delete.

**Comment by navi86** [comment]
@sethvargo thanks you about pointing out on different usage for John
with permission:
`Deleting refs older than 2022-04-08T20:43:56Z on 2 repo(s)...

europe-west2-docker.pkg.dev/project_id/alpine
  ✗ no refs were deleted

europe-west2-docker.pkg.dev/project_id/alpine/alpine
  ✓ latest
  ✓ sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97
`
without permissions:
`WARNING: Running in dry-run mode - nothing will actually be cleaned!

Deleting refs older than 2022-04-08T20:44:00Z on 1 repo(s)...

europe-west2-docker.pkg.dev/project_id/alpine
  ✗ no refs were deleted
`

**Comment by sethvargo** [comment]
Hi @navi86 

Can you please show the full command you're running for each? I don't see the dry-run warning for the first command, and I don't see the debug log output either. The command should be something like:

```sh
GCRCLEANER_LOG=debug gcr-cleaner -repo europe-west2-docker.pkg.dev/project_id/alpine -recursive -dry-run
```

**Comment by navi86** [comment]
Hi @sethvargo 
debug command and output:
`docker run -it -e GOOGLE_APPLICATION_CREDENTIALS=/root/.config/application_default_credentials_personal.json -e GCRCLEANER_LOG=debug -v /home/docker/.config:/root/.config:ro us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli -repo "europe-west2-docker.pkg.dev/project_id/alpine" -tag-filter-all "[a-u,x-z]|[\d]+?v" -dry-run -recursive
{"message":"using default token resolution for authentication","severity":"DEBUG","time":"2022-04-09T09:49:46Z"}
{"message":"gathering child repositories recursively","severity":"DEBUG","time":"2022-04-09T09:49:46Z"}
WARNING: Running in dry-run mode - nothing will actually be cleaned!

Deleting refs older than 2022-04-09T09:49:46Z on 1 repo(s)...

europe-west2-docker.pkg.dev/project_id/alpine
{"message":"computed repo","repo":"europe-west2-docker.pkg.dev/project_id/alpine","severity":"DEBUG","time":"2022-04-09T09:49:46Z"}
  ✗ no refs were deleted
`

**Comment by sethvargo** [comment]
Is there actually anything inside `europe-west2-docker.pkg.dev/project_id/alpine`? I see the log message about recursive listing, but it only found one repo.

**Comment by navi86** [comment]
@sethvargo yes, tKe a look please here
https://github.com/GoogleCloudPlatform/gcr-cleaner/issues/80#issuecomment-1093350077

**Comment by sethvargo** [comment]
@navi86 can you go look in the Cloud Console? I think you ran the previous command with the wrong flag, so it didn't run in dry-run mode, thus deleted the references. That's why the newer invocations are not seeing additional refs to delete.

**Comment by navi86** [comment]
Hi @sethvargo 
I've repushed docker images, so images are there, I run this command again:
`
docker run -it -e GOOGLE_APPLICATION_CREDENTIALS=/root/.config/application_default_credentials_personal.json -e GCRCLEANER_LOG=debug -v /home/docker/.config:/root/.config:ro us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli -repo "europe-west2-docker.pkg.dev/project_id/alpine" -tag-filter-all "[a-u,x-z]|[\d]+?v" -dry-run -recursive`
{"message":"using default token resolution for authentication","severity":"DEBUG","time":"2022-04-11T14:00:06Z"}
{"message":"gathering child repositories recursively","severity":"DEBUG","time":"2022-04-11T14:00:06Z"}
WARNING: Running in dry-run mode - nothing will actually be cleaned!

Deleting refs older than 2022-04-11T14:00:06Z on 1 repo(s)...

europe-west2-docker.pkg.dev/project_id/alpine
{"message":"computed repo","repo":"europe-west2-docker.pkg.dev/project_id/alpine","severity":"DEBUG","time":"2022-04-11T14:00:06Z"}
  ✗ no refs were deleted
❯ `docker run -it -e GOOGLE_APPLICATION_CREDENTIALS=/root/.config/application_default_credentials_personal.json -e GCRCLEANER_LOG=debug -v /home/docker/.config:/root/.config:ro us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli -repo "europe-west2-docker.pkg.dev/project_id/alpine/alpine" -tag-filter-all "[a-u,x-z]|[\d]+?v" -dry-run -recursive`
{"message":"using default token resolution for authentication","severity":"DEBUG","time":"2022-04-11T14:00:14Z"}
{"message":"gathering child repositories recursively","severity":"DEBUG","time":"2022-04-11T14:00:14Z"}
WARNING: Running in dry-run mode - nothing will actually be cleaned!

Deleting refs older than 2022-04-11T14:00:14Z on 1 repo(s)...

europe-west2-docker.pkg.dev/project_id/alpine/alpine
{"message":"computed repo","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-11T14:00:14Z"}
{"digest":"sha256:a777c9c66ba177ccfea23f2a216ff6721e78a662cd17019488c417135299cd89","message":"processing manifest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tags":["3.15.4"],"time":"2022-04-11T14:00:15Z","uploaded":"2022-04-08T19:12:36Z"}
{"digest":"sha256:a777c9c66ba177ccfea23f2a216ff6721e78a662cd17019488c417135299cd89","message":"should not delete","reason":"no filter matches","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-11T14:00:15Z"}
{"digest":"sha256:1e014f84205d569a5cc3be4e108ca614055f7e21d11928946113ab3f36054801","message":"processing manifest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tags":["3.15.3"],"time":"2022-04-11T14:00:15Z","uploaded":"2022-04-08T19:17:59Z"}
{"digest":"sha256:1e014f84205d569a5cc3be4e108ca614055f7e21d11928946113ab3f36054801","message":"should not delete","reason":"no filter matches","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-11T14:00:15Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"processing manifest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tags":["latest"],"time":"2022-04-11T14:00:15Z","uploaded":"2022-04-08T20:40:35Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"should delete","reason":"matches tag filter","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tag_filter":"all([a-u,x-z]|[\\d]+?v)","time":"2022-04-11T14:00:15Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"deleting tag","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tag":"latest","time":"2022-04-11T14:00:15Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"deleting digest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-11T14:00:15Z"}
  ✓ latest
  ✓ sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97
`
P.S. this user doesn't have role or permission that I mentioned here

**Comment by sethvargo** [comment]
Hi @navi86

Thank you for your patience here. I spent some time digging into this this evening. During my investigation, I found a few small bugs and a fairly large performance problem. I did a pretty large refactor of the recursive repo expansion in https://github.com/GoogleCloudPlatform/gcr-cleaner/pull/82, which should dramatically improve the performance and include more detailed logging about how the recursive lister makes decisions.

I'm still struggling to see how the previous code was behaving as described, but I'm hoping this new release either fixes it or provides more logging output.

Could you please pull the latest release (v0.7.3) please? If the issue persists, please supply the debug logs for both invocations.

**Comment by navi86** [comment]
Hi @sethvargo 
it looks like something stopped working, for instance old image:
`docker run -it --rm -e GOOGLE_APPLICATION_CREDENTIALS=/root/.config/application_default_credentials_personal.json -e GCRCLEANER_LOG=debug -v /home/docker/.config:/root/.config:ro us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli@sha256:783ae7fd80e0ff425cd356ca5937895059c5ba6e251c281d4b20399aeee654db -repo "europe-west2-docker.pkg.dev/project_id/alpine/alpine" -tag-filter-all "[a-u,x-z]|[\d]+?v" -dry-run -recursive`
{"message":"using default token resolution for authentication","severity":"DEBUG","time":"2022-04-12T23:21:19Z"}
{"message":"gathering child repositories recursively","severity":"DEBUG","time":"2022-04-12T23:21:20Z"}
WARNING: Running in dry-run mode - nothing will actually be cleaned!

Deleting refs older than 2022-04-12T23:21:20Z on 1 repo(s)...

europe-west2-docker.pkg.dev/project_id/alpine/alpine
{"message":"computed repo","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-12T23:21:20Z"}
{"digest":"sha256:a777c9c66ba177ccfea23f2a216ff6721e78a662cd17019488c417135299cd89","message":"processing manifest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tags":["3.15.4"],"time":"2022-04-12T23:21:20Z","uploaded":"2022-04-08T19:12:36Z"}
{"digest":"sha256:a777c9c66ba177ccfea23f2a216ff6721e78a662cd17019488c417135299cd89","message":"should not delete","reason":"no filter matches","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-12T23:21:20Z"}
{"digest":"sha256:1e014f84205d569a5cc3be4e108ca614055f7e21d11928946113ab3f36054801","message":"processing manifest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tags":["3.15.3"],"time":"2022-04-12T23:21:20Z","uploaded":"2022-04-08T19:17:59Z"}
{"digest":"sha256:1e014f84205d569a5cc3be4e108ca614055f7e21d11928946113ab3f36054801","message":"should not delete","reason":"no filter matches","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-12T23:21:20Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"processing manifest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tags":["latest"],"time":"2022-04-12T23:21:20Z","uploaded":"2022-04-08T20:40:35Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"should delete","reason":"matches tag filter","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tag_filter":"all([a-u,x-z]|[\\d]+?v)","time":"2022-04-12T23:21:20Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"deleting tag","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","tag":"latest","time":"2022-04-12T23:21:20Z"}
{"digest":"sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97","message":"deleting digest","repo":"europe-west2-docker.pkg.dev/project_id/alpine/alpine","severity":"DEBUG","time":"2022-04-12T23:21:20Z"}
  ✓ latest
  ✓ sha256:73c155696fe65b68696e6ea24088693546ac468b3e14542f23f0efbde289cc97

new docker image 
`docker run -it --rm -e GOOGLE_APPLICATION_CREDENTIALS=/root/.config/application_default_credentials_personal.json -e GCRCLEANER_LOG=debug -v /home/docker/.config:/root/.config:ro us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli -repo "europe-west2-docker.pkg.dev/project_id/alpine/alpine" -tag-filter-all "[a-u,x-z]|[\d]+?v" -dry-run -recursive`
{"message":"using default token resolution for authentication","severity":"DEBUG","time":"2022-04-12T23:23:18Z"}
{"message":"gathering child repositories recursively","severity":"DEBUG","time":"2022-04-12T23:23:18Z"}
{"message":"finding all child repositories","roots":["europe-west2-docker.pkg.dev/project_id/alpine/alpine"],"severity":"DEBUG","time":"2022-04-12T23:23:18Z"}
{"message":"listing child repositories for registry","registry":"europe-west2-docker.pkg.dev","severity":"DEBUG","time":"2022-04-12T23:23:18Z"}
{"message":"found child repositories for registry","registry":"europe-west2-docker.pkg.dev","repos":null,"severity":"DEBUG","time":"2022-04-12T23:23:18Z"}
{"in":["europe-west2-docker.pkg.dev/project_id/alpine/alpine"],"message":"recursively listed child repositories","out":[],"severity":"DEBUG","time":"2022-04-12T23:23:18Z"}
WARNING: Running in dry-run mode - nothing will actually be cleaned!

Deleting refs older than 2022-04-12T23:23:18Z on 0 repo(s)...

so basically new docker image does't like that it's using  "-tag-filter-all"

**Comment by sethvargo** [comment]
Hi @navi86 thank you. I don't think the tag filter has anything to do with it. The new log output clearly shows that the `LIST` call for `europe-west2-docker.pkg.dev` returned an empty list:

```json
{"message":"found child repositories for registry","registry":"europe-west2-docker.pkg.dev","repos":null,"severity":"DEBUG","time":"2022-04-12T23:23:18Z"}
```

What is the output of:

```sh
curl -H "Authorization: Bearer $(gcloud auth print-access-token)" "https://europe-west2-docker.pkg.dev/v2/_catalog?n=100"
```

**Comment by Twb3** [comment]
Hi @sethvargo, I am experiencing a similar issue:

When I run this:
```bash
docker run -e GCRCLEANER_LOG=debug -it us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli \
    -dry-run -recursive -repo us-central1-docker.pkg.dev/project_id/repo_name \
    -token "$(gcloud auth application-default print-access-token)"
```
I get:
```json
{"message":"found child repositories for registry","registry":"us-central1-docker.pkg.dev","repos":null,"severity":"DEBUG","time":"2022-04-14T17:06:43Z"}
```

When I run the curl command I get a list of repositories.  My service account only has the `roles/artifactregistry.repoAdmin` permission.
```bash
curl -H "Authorization: Bearer $(gcloud auth print-access-token)" "https://us-central1-docker.pkg.dev/v2/_catalog?n=100"
```

**Comment by sethvargo** [comment]
@Twb3 do you get the list of repos if you grant the service account `roles/storage.objectViewer`?

@navi86 if that's true, that's either a bug in Artifact Registry or a missing documentation.

**Comment by Twb3** [comment]
@sethvargo Yes when I grant the service account `roles/storage.objectViewer` it returns repositories and `-recursive` now works.

**Comment by sethvargo** [comment]
@Twb3 @navi86 - I'm happy to document this in the README, but it would be helpful if you, as customers, also filed bugs. It seems odd that you would need to grant `roles/storage.objectViewer` so a _different_ API can make LIST calls. It's also strange that the response is [] instead of permission denied. Once you file those, I can try to get them prioritized internally.
